RFS/src/main/java/com/rfs/worker/IndexStep.java

```diff
@@ -0,0 +1,475 @@
+            members.cmsEntry = members.cmsClient.getIndexEntry();
+        }
+
+        @Override
+        public WorkerStep nextStep() {
+            if (members.cmsEntry.isEmpty()) {
+                return new CreateEntry(members);
+            } 
+            
+            CmsEntry.Index currentEntry = members.cmsEntry.get();            
+            switch (currentEntry.status) {
+                case SETUP:
+                    // TODO: This uses the client-side clock to evaluate the lease expiration, when we should
+                    // ideally be using the server-side clock.  Consider this a temporary solution until we find
+                    // out how to use the server-side clock.
+                    long leaseExpiryMillis = Long.parseLong(currentEntry.leaseExpiry);
+                    Instant leaseExpiryInstant = Instant.ofEpochMilli(leaseExpiryMillis);
+                    boolean leaseExpired = leaseExpiryInstant.isBefore(Instant.now());
+
+                    // Don't try to acquire the lease if we're already at the max number of attempts
+                    if (currentEntry.numAttempts >= CmsEntry.Index.MAX_ATTEMPTS && leaseExpired) {
+                        return new ExitPhaseFailed(members, new MaxAttemptsExceeded());
+                    }
+
+                    if (leaseExpired) {
+                        return new AcquireLease(members);
+                    } 
+                    
+                    logger.info("Index Migration entry found, but there's already a valid work lease on it");
+                    return new RandomWait(members);
+
+                case IN_PROGRESS:
+                    return new GetIndicesToMigrate(members);
+                case COMPLETED:
+                    return new ExitPhaseSuccess(members);
+                case FAILED:
+                    return new ExitPhaseFailed(members, new FoundFailedIndexMigration());
+                default:
+                    throw new IllegalStateException("Unexpected metadata migration status: " + currentEntry.status);
+            }
+        }
+    }
+
+    public static class CreateEntry extends Base {
+
+        public CreateEntry(SharedMembers members) {
+            super(members);
+        }
+
+        @Override
+        public void run() {
+            logger.info("Index Migration CMS Entry not found, attempting to create it...");
+            members.cmsEntry = members.cmsClient.createIndexEntry();
+            logger.info("Index Migration CMS Entry created");
+        }
+
+        @Override
+        public WorkerStep nextStep() {
+            // Set up the index work entries if we successfully created the CMS entry; otherwise, circle back to the beginning
+            if (members.cmsEntry.isPresent()) {
+                return new SetupIndexWorkEntries(members);
+            } else {
+                return new GetEntry(members);
+            }
+        }
+    }
+
+    public static class AcquireLease extends Base {
```

Comment:
Does index step need its own version of this class, isn't it the same as all other leases?

Reply:
I'm not sure yet, actually.  Probably? I was going to wait until wrapping the final phase (Shard/document migration) before making that determination rather than get too adventurous with abstraction at this point.  There's been a number of surprises so far about some of the abstractions and I kinda want to keep things straightforward, even if some code is repeated, just to get the entire thing on paper before trying to simplify.

Reply:
Seems like this will be clearer after the majority of the work in this space is done.  However, please consider the folks reviewing your code - more code is more to review, to understand, to discuss.  The sooner we can slim this down the easier it will be to pickup the next areas of the project.